helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -93,42 +187,126 @@ public static ClusterModel generateClusterModel(ResourceControllerDataProvider d
     return new ClusterModel(context, toBeAssignedReplicas, assignableNodes);
   }
 
+  // Filter the replicas map so only the replicas that have been allocated in the existing
+  // assignmentMap remain in the map.
+  private static void removeUnknownReplicas(Map<String, Set<AssignableReplica>> replicaMap,
+      Map<String, ResourceAssignment> assignmentMap) {
+    replicaMap.entrySet().parallelStream().forEach(replicaSetEntry -> {
+      // <partition, <state, instances set>>
+      Map<String, Map<String, Set<String>>> stateInstanceMap =
+          getStateInstanceMap(assignmentMap.get(replicaSetEntry.getKey()));
+      // Iterate the replicas of the resource to find the ones that require reallocating.
+      Iterator<AssignableReplica> replicaIter = replicaSetEntry.getValue().iterator();
+      while (replicaIter.hasNext()) {
+        AssignableReplica replica = replicaIter.next();
+        Set<String> validInstances =
+            stateInstanceMap.getOrDefault(replica.getPartitionName(), Collections.emptyMap())
+                .getOrDefault(replica.getReplicaState(), Collections.emptySet());
+        if (validInstances.isEmpty()) {
+          // Removing by comparing with the baseline assignment.
+          replicaIter.remove();
+        } else {
+          // Cleanup the state map record, so the selected instance won't be picked up again for
+          // the other replica checkup.
+          validInstances.remove(validInstances.iterator().next());
+        }
+      }
+    });
+  }
+
   /**
-   * Generate a cluster model based on the current state output and data cache.
-   * @param dataProvider           The controller's data cache.
-   * @param resourceMap            The full list of the resources to be rebalanced. Note that any
-   *                               resources that are not in this list will be removed from the
-   *                               final assignment.
-   * @param currentStateAssignment The resource assignment built from current state output.
-   * @return A cluster model based on the current state and data cache.
+   * Find the minimum set of replicas that need to be reassigned by comparing with the Baseline
+   * assignment.
+   * A replica needs to be reassigned if one of the following conditions is true:
+   * 1. The assignments in the Baseline and the Best possible assignment are different. And the
+   * assignment in the Baseline is valid. So it is worthwhile to move it.
+   * 2. The assignments is not in the Baseline or the Best possible assignment.
```

Comment:
What? This also doesn't make sense..

Reply:
What do you mean? Details? Sorry but this comment does not make sense to me as well.
Let me try to guess, maybe make "The assignments" to "The partition allocation" would make this paragraph easier for you to understand?

Reply:
Sorry. My comment was incomplete. What's confusing to me here is that
If an assignment is not in the baseline or the best possible assignment, it's ambiguous. Did you mean an assignment is in neither baseline nor best possible? Or if it's missing in any of the two, should it be moved?

If you give a simple scenario, it would be very helpful. I know this sounds nitpicky, but please understand that I am trying to provide feedback from a different perspective - I know that these may sound like dumb questions to the author, but for reviewers/readers who didn't write the code, they will thank you for making documentation clear and legible. So let's try to put ourselves in the position of someone who doesn't have much context and make the code self-explanatory where possible.

Reply:
np, It is a complicated part, so any comments are welcome. I think I should say "either of the following conditions is true".
Unfortunately, even a simple scenario would increase a lot of extra readings here. So let me make the minimum change first.